File: 2025/03/main.py
def first_highest_number(num_str,index1,index2):
    high_i = index1
    for i in range(index1+1,index2):
        if num_str[high_i] == "9":
            return high_i
        if num_str[i] > num_str[high_i]:
            high_i = i
    return high_i

def joltage(digits):
    part1 = 0
    for num_str in map(str.strip, open("input.txt")):
        current,last_i = [],-1
        for i in range(digits):
            last_i = first_highest_number(num_str,last_i+1,len(num_str) - (digits - 1 - i))
            current.append(num_str[last_i])
        part1 += int("".join(current))
    return part1

if __name__ == "__main__":
    print("Part 1:",joltage(2))
    print("Part 2:",joltage(12))

File: 2025/03/main.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

func firstHighestNumber(numStr string, start, end int) int {
	bestIdx := start
	for i := start + 1; i < end; i++ {
		if numStr[bestIdx] == '9' {
			return bestIdx
		}
		if numStr[i] > numStr[bestIdx] {
			bestIdx = i
		}
	}
	return bestIdx
}

func solve(lines []string, digits int) int {
	total := 0
	for _, numStr := range lines {
		result := make([]byte, digits)
		lastIdx := -1

		for i := 0; i < digits; i++ {
			lastIdx = firstHighestNumber(numStr, lastIdx+1, len(numStr)-(digits-1-i))
			result[i] = numStr[lastIdx]
		}

		num, _ := strconv.Atoi(string(result))
		total += num
	}
	return total
}

func main() {
	data, _ := os.ReadFile("input.txt")
	lines := strings.Split(strings.TrimSpace(string(data)), "\n")

	fmt.Println("Part 1:", solve(lines, 2))
	fmt.Println("Part 2:", solve(lines, 12))
}
